File: jarvis-desktop/tests/own-network.mjs
/**
 * The Jarvis address: the owner's own networks only (CLAUDE.md, decided
 * 2026-09-26).
 *
 * Settings used to accept any address of the right shape, so a public
 * tunnel (`https://abc.ngrok-free.app`) was saved, and the pairing token went
 * through it with every request. Now an address off the owner's own
 * networks - this PC, the home network, Tailscale, NordVPN Meshnet - is
 * refused in one plain sentence, and one saved before the rule is not used:
 * the page says so in red, and the link line says it the way it says any
 * other connection error.
 *
 * The verdicts and the sentence come from the shared case table,
 * tests/fixtures/own-network-cases.json, which tools/gen_own_network_cases.py
 * writes from the backend's own rule (backend/jarvis_local_http.py). The Rust
 * that applies it is tested against the same table in commands.rs
 * (own_network_tests), and the phone's OwnNetworkTest reads a byte-identical
 * copy.
 */
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import * as K from "./uikit.mjs";

const HERE = dirname(fileURLToPath(import.meta.url));
const read = (p) => readFileSync(join(HERE, "..", p), "utf8");
const CASES = JSON.parse(read("tests/fixtures/own-network-cases.json"));
const say = (address) => CASES.message.replace("{address}", address);
const REFUSED = CASES.origins.filter((c) => !c.own).map((c) => c.url);
const ALLOWED = CASES.origins.filter((c) => c.own).map((c) => c.url);

const fails = [];
const check = async (name, fn) => {
  try { await fn(); console.log(`ok    ${name}`); }
  catch (e) { fails.push(name); console.log(`FAIL  ${name}\n      ${e.message}`); }
};

/* ── No browser: the table, the words, the link line ─────────────────────── */

await check("the table is the one the phone reads, byte for byte", async () => {
  const phone = readFileSync(join(HERE, "..", "..", "jarvis-client", "app", "src", "test",
    "resources", "contract", "own-network-cases.json"), "utf8");
  assert.equal(phone, read("tests/fixtures/own-network-cases.json"));
});

await check("the table holds the tricky cases the rule exists for", async () => {
  const verdict = Object.fromEntries(CASES.hosts.map((c) => [c.host, c.own]));
  for (const h of ["100.63.255.255", "100.128.0.0", "172.15.255.255", "172.32.0.0", "0.0.0.0",
    "localhost.evil.com", "10.0.0.1.nip.io", "abc123.ngrok-free.app",
    "my-jarvis.trycloudflare.com", "169.254.1.1", "fe80::1", "8.8.8.8"]) {
    assert.equal(verdict[h], false, `${h} should be refused`);
  }
  for (const h of ["localhost", "127.0.0.1", "::1", "192.168.1.20", "10.0.0.1", "172.16.0.1",
    "100.64.0.0", "100.127.255.255", "homeassistant.local", "desktop.tail1234.ts.net",
    "marioirelan11-alps.nord", "DESKTOP.TAIL1234.TS.NET", "desktop.tail1234.ts.net."]) {
    assert.equal(verdict[h], true, `${h} should be allowed`);
  }
  assert.ok(REFUSED.some((u) => u.startsWith("https://")), "https:// is refused too");
  assert.ok(CASES.tricky.some((c) => c.url.includes("@") && !c.own), "a user-name trick");
});

await check("the refusal is ONE sentence, so the link line shows all of it", async () => {
  const link = await import("../src/jarvis-link.js");
  const url = "https://abc123.ngrok-free.app";
  const words = link.linkWords({ connected: false, stale: true, error: say(url) });
  assert.equal(words.tone, "bad");
  assert.equal(words.canAct, false, "approving must be blocked");
  // linkWords shows the reason's first sentence; all of it must survive.
  assert.ok(words.text.startsWith(`Offline — Jarvis's address ${url} is not on your own networks`),
    words.text);
  assert.match(words.text, /a name ending in \.nord\)\. Approving is blocked/, words.text);
});

/* ── Settings ────────────────────────────────────────────────────────────── */

const { base, close } = await K.serve();
const browser = await K.launch();
const VIEW = { width: 760, height: 1400 };
const open = (data = {}) => K.open(browser, base, "settings.html", data, VIEW);
const refusals = Object.fromEntries(REFUSED.map((u) => [u, say(u)]));

await check("the field says which addresses are accepted, and why", async () => {
  const page = await open();
  const note = await page.locator("#base-note").innerText();
  await page.close();
  assert.match(note, /own networks/);
  assert.match(note, /Tailscale/);
  assert.match(note, /NordVPN Meshnet/);
  assert.match(note, /ngrok/);
});

for (const url of ["https://abc123.ngrok-free.app", "http://100.128.0.0:4719",
  "http://localhost.evil.com:4719"]) {
  await check(`saving ${url} is refused in the shared words`, async () => {
    const page = await open({ baseRefusals: refusals });
    await page.locator("#base").fill(url);
    await page.locator("#save-connection").click();
    await page.waitForTimeout(250);
    const status = await page.locator("#connection-status").innerText();
    const tone = await page.locator("#connection-status").getAttribute("data-tone");
    await page.close();
    assert.equal(status.trim(), say(url));
    assert.equal(tone, "bad");
  });
}

await check("CONTROL: an address on the owner's own networks saves", async () => {
  const page = await open({ baseRefusals: refusals });
  await page.locator("#base").fill(ALLOWED.find((u) => u.includes(".ts.net")));
  await page.locator("#save-connection").click();
  await page.waitForTimeout(250);
  const tone = await page.locator("#connection-status").getAttribute("data-tone");
  const status = await page.locator("#connection-status").innerText();
  await page.close();
  assert.notEqual(tone, "bad", status);
  assert.match(status, /^Saved/);
});

await check("an address saved before the rule is flagged in red, and said to be unused", async () => {
  const url = "https://abc123.ngrok-free.app";
  const page = await open({ apiSettings: { base: url, baseProblem: say(url) } });
  await page.waitForTimeout(250);
  const field = await page.locator("#base").inputValue();
  const status = await page.locator("#connection-status").innerText();
  const tone = await page.locator("#connection-status").getAttribute("data-tone");
  await page.close();
  assert.equal(field, url, "the field must show the refused address, so it can be changed");
  assert.match(status, /not being used/);
  assert.ok(status.includes(say(url)), status);
  assert.match(status, /Nothing is sent to it/);
  assert.equal(tone, "bad");
});

await browser.close();
close();

/* ── Controls: the Rust that does the refusing ───────────────────────────── */

await check("CONTROL: validate_base asks the own-network rule, tested on the shared table", async () => {
  const rust = read("src-tauri/src/commands.rs");
  const fn = rust.slice(rust.indexOf("fn validate_base"));
  assert.match(fn.slice(0, fn.indexOf("\n}\n")), /own_network_problem\(base\)/);
  assert.match(rust, /include_str!\("\.\.\/\.\.\/tests\/fixtures\/own-network-cases\.json"\)/);
});

await check("CONTROL: a refused configured address is never used, and the stream says why", async () => {
  const rust = read("src-tauri/src/commands.rs");
  const fn = rust.slice(rust.indexOf("pub(crate) fn base_from"));
  const body = fn.slice(0, fn.indexOf("\n}\n"));
  assert.match(rust, /pub fn jarvis_base\(app: &AppHandle\) -> String \{\s*base_from\(configured_base\(app\)\)\s*\}/);
  assert.match(body, /validate_base\(&base\)\.is_ok\(\)/, "jarvis_base does not check");
  // The owner's decision of 2026-09-26: nothing over the network while a
  // refused address is saved - not to it, and not to this PC instead.
  assert.match(body, /Some\(_\) => String::new\(\)/, "a refused address falls back to something");
  assert.match(body, /None => DEFAULT_BASE/, "an unset address no longer means this PC");
  const headers = rust.slice(rust.indexOf("pub fn jarvis_headers("));
  assert.match(headers.slice(0, headers.indexOf("\n}\n")), /require_base_allowed\(app\)\?;/,
    "requests to Jarvis are not refused while a refused address is saved");
  const stream = read("src-tauri/src/stream.rs");
  const loop = stream.slice(stream.indexOf("pub fn spawn"));
  const problem = loop.indexOf("commands::base_problem(&app)");
  const connect = loop.indexOf("connect_once(");
  assert.ok(problem > -1 && problem < connect, "the stream connects before checking");
});

if (fails.length) {
  console.log(`\n${fails.length} failed`);
  process.exit(1);
}
console.log("\nall passed");

File: jarvis-client/app/src/main/java/com/jarvis/client/data/OwnNetwork.kt
package com.jarvis.client.data

import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * The desktop address must be on the owner's own networks (CLAUDE.md,
 * decided 2026-09-26): this PC, the home network, Tailscale or NordVPN
 * Meshnet. Anything on the open internet - a public tunnel such as ngrok or
 * Cloudflare included - is refused, so the pairing token never goes there.
 * https:// is refused off those networks too: the question is where the
 * token goes, not whether the line is scrambled.
 *
 * NOT a rule of this app's own. It is the backend's, word for word:
 * `backend/jarvis_local_http.py`, `_own_network` - the rule that already
 * decides where plain http:// may carry the Home Assistant token or the
 * calendar password. tools/gen_own_network_cases.py runs that real code over
 * a table of cases and writes `contract/own-network-cases.json`, which
 * OwnNetworkTest reads; the desktop's Rust (commands.rs, own_network_tests)
 * reads a byte-identical copy, so the two apps cannot drift apart unnoticed.
 *
 * Judged by spelling alone: nothing is looked up, no DNS, no network call.
 *
 * Where it is applied: [ClientSettings.baseUrl] gives no address at all for
 * a refused one, so nothing is ever sent there, and every request and the
 * event stream say [problem]'s sentence instead (JarvisApi, EventStream) -
 * the way any other connection failure is shown.
 *
 * Pure Kotlin, no Android types, so it runs on a plain JVM. OkHttp's
 * [toHttpUrlOrNull] is plain JVM code too - and it is the parser the
 * requests themselves use, which is why it is asked (see [problem]).
 */
object OwnNetwork {

    /**
     * What the owner sees when an address is refused - one sentence, checked
     * against `phone_message` in the shared table. `{address}` is the address
     * as saved. Its first half is the desktop's; its ending names only what
     * this phone can connect to: network_security_config.xml allows plain
     * http:// only to .ts.net and .nord names (and the phone itself), so a
     * home-network address that passes [problem] still cannot be reached
     * (ease-of-use audit 2026-09-27, #1e; PlatformReadiness says so for one).
     */
    const val MESSAGE =
        "Jarvis's address {address} is not on your own networks, so this app will not " +
            "send your pairing key there: on this phone, use your PC's Tailscale name " +
            "(ending in .ts.net) or its NordVPN Meshnet name (ending in .nord)."

    /** Name endings only the owner's own networks answer - the backend's `_OWN_SUFFIXES`. */
    private val OWN_SUFFIXES = listOf(".local", ".lan", ".home.arpa", ".ts.net", ".nord")

    /**
     * Null when [base] (a whole address with its scheme, as [BaseUrl.normalise]
     * makes it) is on the owner's own networks, else the sentence saying why
     * it is refused.
     *
     * Two readings of the host must BOTH pass, the way the backend judges
     * both `urlsplit`'s host and the one urllib dials: a plain split of the
     * text as written, and the host OkHttp parses - the parser every request
     * uses, which also decodes `%6c`-escapes and would take `a@b` as a user
     * name. So an address one of them misreads cannot slip past the other.
     */
    fun problem(base: String): String? {
        val b = base.trim().trimEnd('/')
        val typed = typedHost(b)
        val dialled = runCatching { b.toHttpUrlOrNull()?.host }.getOrNull()
        if (typed != null && dialled != null && ownHost(typed) && ownHost(dialled)) return null
        return message(b)
    }

    /**
     * [MESSAGE] for [base], which is shown as written - unless it holds
     * something that must not be repeated back (a user name or password
     * written into it, "me:pw@...", or a space), when the sentence leaves the
     * address out.
     */
    fun message(base: String): String {
        val unshown = base.isEmpty() || base.any { it == '@' || it.isWhitespace() || it.isISOControl() }
        return if (unshown) MESSAGE.replace("{address} ", "") else MESSAGE.replace("{address}", base)
    }

    /**
     * The host as the address is written: after `scheme://`, up to the first
     * `/`, `?`, `#` or `\`, without a `:port`, and without IPv6's brackets.
     * Null when that is not one host (an unbracketed IPv6 address, which
     * reads as a host and a port, or a port that is not a number).
     */
    internal fun typedHost(base: String): String? {
        val rest = if ("://" in base) base.substringAfter("://") else base
        val authority = rest.takeWhile { it != '/' && it != '?' && it != '#' && it != '\\' }
        val host: String
        val port: String?
        if (authority.startsWith("[")) {
            val close = authority.indexOf(']')
            if (close < 0) return null
            host = authority.substring(1, close)
            val after = authority.substring(close + 1)
            port = when {
                after.isEmpty() -> null
                after.startsWith(":") -> after.substring(1)
                else -> return null
            }
        } else {
            host = authority.substringBefore(':')
            port = if (':' in authority) authority.substringAfter(':') else null
        }
        if (port != null && port.isNotEmpty() && !port.all { it in '0'..'9' }) return null
        return host.ifEmpty { null }
    }

    /**
     * Is [host] this PC or on one of the owner's own networks? The backend's
     * `_own_network`, line for line:
     *
     * - an address (in any spelling the operating system would still dial -
     *   `3232235777`, `0xc0a80101`, `10.1` - judged as that address) must be
     *   in 127.0.0.0/8, ::1, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16,
     *   fc00::/7 (which holds Tailscale's fd7a:115c:a1e0::/48) or
     *   100.64.0.0/10 (Tailscale and NordVPN Meshnet). Link-local
     *   (169.254.x.x, fe80::) is not on the list, the same as the backend;
     * - a name must be `localhost`, a single word with no dot (a
     *   home-network name such as `nas`, which the home router answers), or
     *   end in `.local`, `.lan`, `.home.arpa`, `.ts.net` or `.nord`.
     */
    fun ownHost(host: String): Boolean {
        val h = host.trim().lowercase().trimEnd('.')
        if (h.isEmpty()) return false
        address(h)?.let { return ownAddress(it) }
        if (!plainName(h)) return false // an odd IPv6 form, an "@", a space, a "%"...
        if (h == "localhost" || '.' !in h) return true
        return OWN_SUFFIXES.any { h.endsWith(it) }
    }

    /** The backend's `_OWN_NETS`, for a 4- or 16-byte address. */
    private fun ownAddress(ip: ByteArray): Boolean {
        val a = ip[0].toInt() and 0xff
        val b = ip[1].toInt() and 0xff
        if (ip.size == 4) {
            return a == 127 || a == 10 || (a == 172 && b in 16..31) ||
                (a == 192 && b == 168) || (a == 100 && b in 64..127)
        }
        val loopback = (0 until 15).all { ip[it].toInt() == 0 } && ip[15].toInt() == 1
        return loopback || (a and 0xfe) == 0xfc
    }

    /**
     * The address [host] (lower case) denotes, 4 or 16 bytes, or null when it
     * is a name - the backend's `_as_address`: an IPv6 address, else the
     * numeric forms `inet_aton` reads. An IPv4-mapped IPv6 address
     * (`::ffff:192.168.1.1`) is judged as its IPv4 address.
     */
    internal fun address(host: String): ByteArray? {
        val v6 = ipv6(host)
        if (v6 != null) {
            val mapped = (0 until 10).all { v6[it].toInt() == 0 } &&
                (v6[10].toInt() and 0xff) == 0xff && (v6[11].toInt() and 0xff) == 0xff
            return if (mapped) v6.copyOfRange(12, 16) else v6
        }
        return inetAton(host)
    }

    /**
     * The classic BSD `inet_aton` reading of a numeric host: one to four
     * parts separated by dots, each decimal, `0x` hex or leading-zero octal,
     * the last part filling whatever bytes are left. Each part must start
     * with a digit and hold only hex digits or `x`, as glibc's does (so a
     * `+` or `-` sign is never read as part of a number).
     */
    private fun inetAton(s: String): ByteArray? {
        val parts = s.split('.')
        if (parts.size > 4) return null
        val numeric = parts.all { p ->
            p.isNotEmpty() && p[0] in '0'..'9' && p.all { it in '0'..'9' || it in 'a'..'f' || it == 'x' }
        }
        if (!numeric) return null
        val values = parts.map { p ->
            val (digits, radix) = when {
                p.startsWith("0x") -> p.substring(2) to 16
                p.length > 1 && p.startsWith("0") -> p.substring(1) to 8
                else -> p to 10
            }
            if (digits.isEmpty()) return null
            digits.toLongOrNull(radix) ?: return null
        }
        val head = values.dropLast(1)
        val last = values.last()
        if (head.any { it > 0xff }) return null
        val tailBits = 8 * (4 - head.size)
        if (last >= (1L shl tailBits)) return null
        var out = last
        head.forEachIndexed { i, v -> out = out or (v shl (24 - 8 * i)) }
        return byteArrayOf(
            (out shr 24).toByte(), (out shr 16).toByte(), (out shr 8).toByte(), out.toByte(),
        )
    }

    /** A strict dotted quad: four decimal parts, no leading zeros, each at most 255. */
    private fun strictIpv4(s: String): ByteArray? {
        val parts = s.split('.')
        if (parts.size != 4) return null
        val bytes = ByteArray(4)
        parts.forEachIndexed { i, p ->
            if (p.isEmpty() || p.length > 3 || !p.all { it in '0'..'9' }) return null
            if (p.length > 1 && p[0] == '0') return null
            val v = p.toInt()
            if (v > 255) return null
            bytes[i] = v.toByte()
        }
        return bytes
    }

    /**
     * A textual IPv6 address (lower case, no brackets, no zone) as 16 bytes,
     * or null. Groups of one to four hex digits; one `::` standing for at
     * least one group of zeros; an IPv4 dotted quad allowed as the last 32
     * bits.
     */
    private fun ipv6(s: String): ByteArray? {
        if (':' !in s) return null
        if (!s.all { it in '0'..'9' || it in 'a'..'f' || it == ':' || it == '.' }) return null
        var text = s
        var v4: ByteArray? = null
        val lastColon = text.lastIndexOf(':')
        if ('.' in text.substring(lastColon + 1)) {
            v4 = strictIpv4(text.substring(lastColon + 1)) ?: return null
            text = text.substring(0, lastColon + 1) + "0:0"
        }
        val halves = text.split("::")
        if (halves.size > 2) return null
        val head = groups(halves[0]) ?: return null
        val tail = if (halves.size == 2) groups(halves[1]) ?: return null else emptyList()
        val count = head.size + tail.size
        if (halves.size == 1 && count != 8) return null
        if (halves.size == 2 && count > 7) return null
        val all = head + List(8 - count) { 0 } + tail
        val bytes = ByteArray(16)
        all.forEachIndexed { i, g ->
            bytes[2 * i] = (g shr 8).toByte()
            bytes[2 * i + 1] = g.toByte()
        }
        v4?.copyInto(bytes, destinationOffset = 12)
        return bytes
    }

    /** "1:ab:0" as [1, 0xab, 0]; "" as none; null for an empty or over-long group. */
    private fun groups(part: String): List<Int>? {
        if (part.isEmpty()) return emptyList()
        return part.split(':').map { g ->
            if (g.isEmpty() || g.length > 4) return null
            g.toIntOrNull(16) ?: return null
        }
    }

    /**
     * The backend's `_NAME_RE`, `^[\w-]+(\.[\w-]+)*$`: words of letters,
     * digits, `_` or `-`, joined by single dots.
     */
    private fun plainName(host: String): Boolean =
        host.split('.').all { label ->
            label.isNotEmpty() && label.all { it.isLetterOrDigit() || it == '_' || it == '-' }
        }
}
